server/application-server/src/main/java/de/tum/in/www1/hephaestus/leaderboard/SlackMessageService.java

```diff
@@ -0,0 +1,148 @@
+package de.tum.in.www1.hephaestus.leaderboard;
+
+import java.io.IOException;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.IntStream;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.context.event.ApplicationReadyEvent;
+import org.springframework.context.event.EventListener;
+import org.springframework.core.Ordered;
+import org.springframework.core.annotation.Order;
+import org.springframework.scheduling.annotation.EnableScheduling;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+
+import com.slack.api.bolt.App;
+import com.slack.api.methods.SlackApiException;
+import com.slack.api.methods.request.chat.ChatPostMessageRequest;
+import com.slack.api.methods.response.auth.AuthTestResponse;
+import com.slack.api.methods.response.chat.ChatPostMessageResponse;
+import com.slack.api.model.User;
+import com.slack.api.model.block.LayoutBlock;
+
+import static com.slack.api.model.block.Blocks.*;
+import static com.slack.api.model.block.composition.BlockCompositions.*;
+import static com.slack.api.model.block.element.BlockElements.*;
+
+@Order(value = Ordered.LOWEST_PRECEDENCE)
+@EnableScheduling
+@Service
+public class SlackMessageService {
+
+    private static final Logger logger = LoggerFactory.getLogger(SlackMessageService.class);
+
+    @Value("${slack.channelId}")
+    private String channelId;
+
+    @Value("${slack.runScheduledMessage}")
+    private boolean runScheduledMessage;
+
+    @Autowired
+    private App slackApp;
+
+    @Autowired
+    private LeaderboardService leaderboardService;
+
+    public void sendMessage(String channelId, List<LayoutBlock> blocks, String fallback)
+            throws IOException, SlackApiException {
+        ChatPostMessageRequest request = ChatPostMessageRequest.builder()
+                .channel(channelId)
+                .blocks(blocks)
+                .text(fallback) // used for example in notifications
+                .build();
+
+        ChatPostMessageResponse response = slackApp.client().chatPostMessage(request);
+
+        if (!response.isOk()) {
+            logger.error("Failed to send message to Slack channel: " + response.getError());
+        }
+    }
+
+    private List<User> getTop3SlackReviewers() {
+        LocalDate after = LocalDate.now().minusDays(8);
+        LocalDate before = LocalDate.now().minusDays(2);
```

Comment:
I'm still unsure about this 🤔  IMO you need to include everything from post to post, or at least Tuesday to Monday, also I think you are hardcoding stuff again with this minus 2 days for Tuesday, minus 1 seemed more correct. 

We could potentially include the 9:00am timestamp in the leaderboard query so it really goes from Tuesday 9:00am to Tuesday 9:00am. We will have to think about how to select the timeframes in the filter.

Reply:
So as mentioned in the other comment, we could definitely do this. To make processing results easier, I would also prefer to store leaderboard results directly in the DB, so it's easier to build gamification on top of it.